File: projects/m5/001-budget-app/solutions/federicosiddi/budget.py
class Category:
    def __init__(self, name):
        self.ledger = []
        self.name = name
        self.expenses = 0
    
    def __str__(self):
        output = self.name.center(30, '*') + "\n"
        for entry in self.ledger:
            if len(entry['description']) > 23:
                output += entry['description'][:23]
            else:
                output += entry['description'].ljust(23) 
            output += ("{0:.2f}".format(entry['amount'])).rjust(7) + "\n"
        balance = self.get_balance()
        output += f"Total: {balance}"
        return output

    def deposit(self, amount, description = ""):
        item = {"amount": amount, "description":description}
        self.ledger.append(item)
    
    def withdraw(self, amount, description = ""):
        if self.check_funds(amount):
            item = {"amount": -amount, "description": description}
            self.ledger.append(item)
            return True
        return False
    
    def get_balance(self):
        balance = 0
        for entry in self.ledger:
            balance += float(entry["amount"])
        return balance
    
    def transfer(self, amount, category):
        if self.check_funds(amount):
            self.withdraw(amount, f"Transfer to {category.name}")
            category.deposit(amount, f"Transfer from {self.name}")
            return True
        return False
    
    def check_funds(self, amount):
        if amount > self.get_balance():
            return False
        return True



def create_spend_chart(categories):
    output = "Percentage spent by category\n"
    money_spent_per_cat = {}
    
    for cat in categories:
        for entry in cat.ledger:
            if entry['amount']  < 0 and ("Transfer") not in entry['description']:
                cat.expenses += abs(entry['amount'])
        money_spent_per_cat[cat.name] = cat.expenses
    tot_expenses = sum(money_spent_per_cat.values())
    exp_percentages = {cat: round(exp/tot_expenses*100) for cat, exp in money_spent_per_cat.items()}

    for x in range (100, -10, -10):
        output += str(x).ljust(3)+"| "
        for entry in categories:
            if exp_percentages[entry.name]+5 >= x:
                output += "o  "
            else:
                output += "   "
        output += "\n"
    output += (" ")*4 + ("---")*len(categories)+ "-" + "\n"
    
    for x in range(max(map(len, exp_percentages))):
        output += "    "
        for name in exp_percentages.keys():
            if len(name) > x:
                output += f" {name[x]} "
            else:
                output += "   "
        output += "\n"
    return output

# TEST
if __name__ == "__main__":
    food = Category("Food")
    food.deposit(1000, "initial deposit")
    food.withdraw(10.15, "groceries")
    food.withdraw(15.89, "restaurant and more food for dessert")
    print(food.get_balance())
    clothing = Category("Clothing")
    food.transfer(50, clothing)
    clothing.withdraw(25.55)
    clothing.withdraw(100)
    auto = Category("Auto")
    auto.deposit(1000, "initial deposit")
    auto.withdraw(15)

    print(food)
    print(clothing)

    print(create_spend_chart([food, clothing, auto]))
    

File: projects/m5/001-budget-app/solutions/filipporoccamatisi/budget.js
const util = require("util");

class Category {
    constructor(name) {
        this.name = name;
        this.ledger = [];
    }
    // formats the object when it is printed
    [util.inspect.custom]() {
        // title :
        // *************title*************
        let title = this.name;
        // number of * initials and endings
        const initialPadding = Math.floor((30 - title.length) / 2);
        const finalPadding = Math.round((30 - title.length) / 2);
        // add ** to title
        title = title.padStart(title.length + initialPadding, "*");
        title = title.padEnd(title.length + finalPadding, "*");

        // body
        let body = "";
        for (let mov of this.ledger) {
            let formattedDescription = mov.description;
            // the amount must have two decimal places
            let formattedAmount = mov.amount.toFixed(2);
            // the description must be less than 23 characters
            if (mov.description.length > 23) {
                formattedDescription = mov.description.slice(0, 23)
            }
            // the number of digits must be less than 7
            if (String(mov.amount).length > 7) {
                formattedAmount = Number(mov.amount.toFixed(0).slice(0, 5)).toFixed(2);
            }
            // create and format row : 
            // description       amount
            // description       amount
            let row = formattedDescription;
            row = row.padEnd(30 - formattedAmount.length, " ");
            row = row + formattedAmount;
            body += row + '\n';
        }
        // total
        const total = `Total: ${this.get_balance().toFixed(2)}`;
        // create the complete chart
        return title + '\n' + body + total;
    }
    deposit(amount, description = "") {
        this.ledger.push({
            amount,
            description
        });
    }
    withdraw(amount, description = "") {
        // if there are not enough funds, return false
        if (!this.check_funds(amount)) return false;
        const formattedAmount = amount * -1;
        this.ledger.push({
            amount: formattedAmount,
            description
        });
        return true;
    }
    // returns the current balance of the budget category based 
    // on the deposits and withdrawals that have occurred.
    get_balance() {
        const balance = this.ledger.reduce((reducer, movement) => reducer + movement.amount, 0);
        return balance;
    }
    // transfers funds between two categories
    transfer(amount, category) {
        if (!this.check_funds(amount)) return false;
        this.withdraw(amount, `Transfer to ${category.name}`);
        category.deposit(amount, `Transfer from ${this.name}`);
        return true;
    }
    //checks that there are sufficient
    // funds to withdraw the amount
    check_funds(amount) {
        return this.get_balance() >= amount
    }
}


const roundDownToNearest10 = function (num) {
    return Math.floor(num / 10) * 10;
}

const create_spend_chart = function (listOfCategories) {
    const expenses = [];
    const listOfPercentageSpent = [];
    // sum of all expenses
    let sumOfwithdrawals = 0

    for (let category of listOfCategories) {
        // sum of the expenses of the category
        let categoryExpensesSum = 0
        for (let mov of category.ledger) {
            // if the move is less than 0 then it is an expense
            if (mov.amount < 0) {
                categoryExpensesSum += mov.amount;
            }
        }
        sumOfwithdrawals += categoryExpensesSum;
        expenses.push({ amount: categoryExpensesSum, category: category.name });
    };

    for (let categoryExpenses of expenses) {
        // calculate the percentage spent
        const percentageSpent = ((categoryExpenses.amount / sumOfwithdrawals) * 100)
        // round down to the nearest 10.
        const formattedPercentageSpent = roundDownToNearest10(percentageSpent) || 0;
        // add to list
        listOfPercentageSpent.push({ percentageSpent: formattedPercentageSpent, category: categoryExpenses.category });
    }

    // create chart 
    let chart = "";
    // add 100 - 90 - 80 ecc to the chart
    for (let percentage = 100; percentage >= 0; percentage -= 10) {
        if (percentage === 100) {
            chart += percentage + "|";
        }
        else if (percentage > 0) {
            chart += " " + percentage + "|";
        }
        else {
            chart += "  " + percentage + "|";
        }
        // add O to chart 
        for (let i = 0; i < listOfPercentageSpent.length; i++) {
            if (percentage <= listOfPercentageSpent[i].percentageSpent) {
                chart += " O"
            } else {
                chart += "  "
            }
        }
        chart += '\n';
    }
    // add ----- to the chart
    let horizontalLine = "    ";
    chart += horizontalLine.padEnd((3 * listOfCategories.length) + 4, "-");
    chart += '\n     ';

    // add the category names to the chart
    let maxLength = 0;
    for (let { category } of listOfPercentageSpent) {
        if (category.length > maxLength) maxLength = category.length;
    }
    for (let i = 0; i < maxLength; i++) {
        for (let { category } of listOfPercentageSpent) {
            if (category[i]) {
                chart += category[i] + " ";
            } else {
                chart += "  "
            }
        };
        chart += '\n     ';
    }
    return chart;
}

const food = new Category("Food")
food.deposit(1000, "initial deposit")
food.withdraw(10.15, "groceries")
food.withdraw(15.89, "restaurant and more food for dessert")
console.log(food.get_balance())

const clothing = new Category("Clothing")
food.transfer(50, clothing)
clothing.withdraw(25.55)
clothing.withdraw(100)

const auto = new Category("Auto")
auto.deposit(1000, "initial deposit")
auto.withdraw(15)

console.log(food)
console.log(clothing)
console.log(food.ledger);
console.log(clothing.ledger);
console.log(auto.ledger);


console.log(create_spend_chart([food, clothing, auto]))